### PR TITLE
WASM.SIMD BitSelect & Shuffle ops + Tests

### PR DESCRIPTION
**Pull request by Krovatkin**
This PR adds 
 * BitSelect + tests
 * Shuffle + tests

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/3763)

### CODE REVIEW

**Comment by Krovatkin** [comment]
@Cellule could you please take a look when you have a mnt? Thanks!

**Comment by Krovatkin** [comment]
ping @Cellule

**Comment by Krovatkin** [comment]
@Cellule ping.

**Comment by Cellule** [comment]
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 3986 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KutxdVUzUxO9EpmY8Wb:-KutxdVUzUxO9EpmY8Wc:bb0q6qi) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Backend/IRBuilderAsmJs.cpp#L3986)):*
> ```cpp
> IRBuilderAsmJs::BuildInt32x4_4(Js::OpCodeAsmJs newOpcode, uint32 offset, BUILD_SIMD_ARGS_REG4)
> {
>     ValueType valueType = GetSimdValueTypeFromIRType(TySimd128F4);
> ```

Shouldn't the IRType be `TySimd128I4` in this function ?

---

*[lib/Backend/IRBuilderAsmJs.cpp, line 5843 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-Kuu-v6aRaZ59pCyI7aa:-Kuu-v6aRaZ59pCyI7ab:b4xa3cg) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Backend/IRBuilderAsmJs.cpp#L5843)):*
> ```cpp
>     IR::RegOpnd * src2Opnd = BuildSrcOpnd(GetRegSlotFromSimd128Reg(layout->R2), TySimd128U16);
> 
>     IR::RegOpnd * src3Opnd =  (IR::RegOpnd*)IR::IntConstOpnd::New(layout->INDICES[0], TyInt32, this->m_func);
> ```

Could we make a loop for these and `AddExtendedArg` instead of copy/pasting 16 times

---

*[lib/Runtime/ByteCode/AsmJsByteCodeWriter.cpp, line 256 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuPwZkCzgHTLBrD8c0:-KuuPwZkCzgHTLBrD8c1:bixurj4) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Runtime/ByteCode/AsmJsByteCodeWriter.cpp#L256)):*
> ```cpp
>     bool AsmJsByteCodeWriter::TryWriteAsmShuffle(OpCodeAsmJs op, RegSlot R0, RegSlot R1, RegSlot R2, uint8 indices[])
>     {
>         const uint32 MAX_LANES = 16;
> ```

Can we use `Wasm::Simd::MAX_LANES` instead ?
Same on other places where we have an hardcoded `16`

---

*[lib/Runtime/ByteCode/OpCodesSimd.h, line 432 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KutzGokOSDQezO8eMod:-KutzGokOSDQezO8eMoe:b-xzpio8) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Runtime/ByteCode/OpCodesSimd.h#L432)):*
> V8

Shouldn't this be `U8` ?

---

*[lib/Runtime/Language/InterpreterProcessOpCodeAsmJs.h, line 2187 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuP68G_Y-qKBpVlEom:-KuuP68HamLWvTZ7x-Dc:bvb8dhg) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Runtime/Language/InterpreterProcessOpCodeAsmJs.h#L2187)):*
> ```objc
> 
> // v8x16shuffle
> #define PROCESS_SIMD_V8X16_2I16toV8X16_1_COMMON(name, func, suffix) \
> ```

This is very specific, could we just make this a CUSTOM process and use a helper instead.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 605 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuPPv6fmdUSKHXf2lm:-KuuPPv6fmdUSKHXf2ln:bkzlpjl) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/WasmReader/WasmBinaryReader.cpp#L605)):*
> ```cpp
> void WasmBinaryReader::ShuffleNode()
> {
>     for (uint32 i = 0; i < Simd::MAX_LANES; i++)
> ```

Add `CheckBytesLeft(Simd::MAX_LANES);`

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1457 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuR1wIl3wMpOcNl4mL:-KuuR1wIl3wMpOcNl4mM:b-r8gtbh) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/WasmReader/WasmByteCodeGenerator.cpp#L1457)):*
> ```cpp
> }
> 
> EmitInfo WasmBytecodeGenerator::EnregisterIntConst(uint constVal)
> ```

I don't see any other uses of this function, do we really need to extract this ?

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1484 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuQn0_qEi_78yMq_Ka:-KuuQn0_qEi_78yMq_Kb:by8xger) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/WasmReader/WasmByteCodeGenerator.cpp#L1484)):*
> ```cpp
> EmitInfo WasmBytecodeGenerator::EmitM128BitSelect()
> {
>     EmitInfo mask = PopEvalStack(WasmTypes::M128, _u("Argument should be of type M128"));
> ```

We can leave the default error message for type mismatch. This message shows less information than the default

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 1510 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KuuQU9jeFU5Dv2CJ_MK:-KuuQU9jeFU5Dv2CJ_ML:b32grqb) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/WasmReader/WasmByteCodeGenerator.cpp#L1510)):*
> ```cpp
>     }
> 
>     m_writer->AsmShuffle(Js::OpCodeAsmJs::Simd128_Shuffle_V8X16, resultReg, arg1Info.location, arg2Info.location, GetReader()->m_currentNode.shuffle.indices);
> ```

nit: just use `indices` here instead of using the reader again

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3763)*

**Comment by Krovatkin** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[lib/Runtime/ByteCode/OpCodesSimd.h, line 432 at r1](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-KutzGokOSDQezO8eMod:-KuuvERuiMDYa1j543Lc:b-hvq577) ([raw file](https://github.com/microsoft/chakracore/blob/7c3ca9ce8f3bc85650a853cebdddb476542ff237/lib/Runtime/ByteCode/OpCodesSimd.h#L432)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

> V8

Shouldn't this be `U8` ?
</blockquote></details>

There's already `Simd128_Shuffle_U16` which uses a different implementation to extract indices. I find `Simd128_Shuffle_V8X16` to be a bit more consistent w/ the corresponding WASM opcode. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3763)*

**Comment by Krovatkin** [comment]
tagging @arunetm

**Comment by Cellule** [comment]
Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/Runtime/Language/InterpreterStackFrame.h, line 264 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-Kuv-vulZxvbtHZXaKfV:-Kuv-vumi2GSWsJ4Vj6J:b4l1jzf) ([raw file](https://github.com/microsoft/chakracore/blob/287b00f6e856e183cfe9088a41ebd98a16007f63/lib/Runtime/Language/InterpreterStackFrame.h#L264)):*
> ```objc
>         template <class T> void OP_SimdUint32x4FromFloat32x4(const unaligned T* playout);
>         template <class T> void OP_WasmSimdConst(const unaligned T* playout);
>         template <class T> inline void OP_SimdShuffleV8X16(const unaligned T* playout);
> ```

nit: `inline` is not necessary

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 609 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-Kuv0S3Aw1GZowNMNuEg:-Kuv0S3Aw1GZowNMNuEh:bkj2f98) ([raw file](https://github.com/microsoft/chakracore/blob/287b00f6e856e183cfe9088a41ebd98a16007f63/lib/WasmReader/WasmBinaryReader.cpp#L609)):*
> ```cpp
>         m_currentNode.shuffle.indices[i] = ReadConst<uint8>();
>     }
>     CheckBytesLeft(Simd::MAX_LANES);
> ```

We should check before reading the byte as it could potentially trigger an A/V 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3763)*

File: lib/Runtime/ByteCode/AsmJsByteCodeWriter.h [hunk 1]
```diff
@@ -51,6 +51,9 @@ namespace Js
         IMP_IWASM void AsmReg9(OpCodeAsmJs op, RegSlot R0, RegSlot R1, RegSlot R2, RegSlot R3, RegSlot R4, RegSlot R5, RegSlot R6, RegSlot R7, RegSlot R8);
         IMP_IWASM void AsmReg17(OpCodeAsmJs op, RegSlot R0, RegSlot R1, RegSlot R2, RegSlot R3, RegSlot R4, RegSlot R5, RegSlot R6, RegSlot R7, RegSlot R8,
             RegSlot R9, RegSlot R10, RegSlot R11, RegSlot R12, RegSlot R13, RegSlot R14, RegSlot R15, RegSlot R16);
+        IMP_IWASM void AsmReg19(OpCodeAsmJs op, RegSlot R0, RegSlot R1, RegSlot R2, RegSlot R3, RegSlot R4, RegSlot R5, RegSlot R6, RegSlot R7, RegSlot R8,
```
Comment (MikeHolman) [inline lib/Runtime/ByteCode/AsmJsByteCodeWriter.h]: why the change with AsmReg19? Seems like this isn't being used in wasm?

**Comment by Krovatkin** [comment]
@Cellule could you please give it another look?

File: lib/Runtime/ByteCode/AsmJsByteCodeWriter.h (hunk 1 above)
Comment (Krovatkin) [reply lib/Runtime/ByteCode/AsmJsByteCodeWriter.h]: Ah, good catch!

**Comment by Krovatkin** [comment]
Review status: 22 of 25 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 609 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-Kuv0S3Aw1GZowNMNuEg:-Kuv5FBtcA6-X-HsQH3b:b-njvbl6) ([raw file](https://github.com/microsoft/chakracore/blob/287b00f6e856e183cfe9088a41ebd98a16007f63/lib/WasmReader/WasmBinaryReader.cpp#L609)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary><blockquote>

We should check before reading the byte as it could potentially trigger an A/V 
</blockquote></details>

oh, my bad. I thought I'd put before the loop

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3763)*

**Comment by Cellule** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/3763#-:-Kuv8O9yr0fd0g8TaeRu:bnfp4nl)*

**Comment by Cellule** [comment]
@dotnet-bot test this please

**Comment by Krovatkin** [comment]
@dotnet-bot test Windows 7 ci_dev12_x64_debug please

**Comment by Krovatkin** [comment]
@Cellule did you do **test this** because **Windows 7 ci_dev12_x64_debug** got stuck or was it another error?

**Comment by Krovatkin** [comment]
@MikeHolman @Cellule , thank you!
